nixos/modules/config/system-path.nix

```diff
@@ -7,6 +7,118 @@ with lib;
 
 let
 
+  # This type can accept three different types of expressions:
+  #   * A list of packages like `[ pkgs.foo pkgs.bar ]`
+  #   * An attribute set of packages like `{ inherit (pkgs) foo bar; }`
+  #   * A more structured attribute set of packages like `{ "foo".package = pkgs.foo.override { hasBar = false; }; }`
+  #
+  # The first two cases will be coerced into the third one, which allows the
+  # package to be disabled and its package and list of outputs to be overridden.
+  packagesType =
+    let
+      inherit (lib) types;
+
+      outputsFor = drv:
+        if drv.outputSpecified or false
+        then [ drv.outputName ]
+        else ((drv.meta or {}).outputsToInstall or [ (drv.outputName or "out") ]);
+
+      mkOutputs =
+        let
+          mkEntry = output: lib.nameValuePair output true;
+        in
+        outputs: lib.listToAttrs (map mkEntry outputs);
+
+      drvToAttrs = drv:
+        let
+          defaultOutput = (drv.${(lib.head (drv.outputs or [ "out" ]))}
+            or (throw "something is up with the outputs of this derivation: ${drv}")) // {
```

Comment:
Same here, this should be deduplicated with a lib function. We can also more easily test the code that way.